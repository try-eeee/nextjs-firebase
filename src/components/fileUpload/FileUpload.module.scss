.previewContainer {
  position: relative;
  display: inline-block;
}

.closeButton {
  position: absolute;
  top: -5px;
  right: -5px;
  background-color: transparent;
  border: none;
  color: red;
  cursor: pointer;
  font-size: 1.2rem;
}

.preview {
  width: 100px;
  height: 100px;
  object-fit: cover;
  border-radius: 50%;
  margin-bottom: 1rem;
}

.inputHidden {
  visibility: hidden;
}
