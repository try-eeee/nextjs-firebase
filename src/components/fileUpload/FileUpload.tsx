/* eslint-disable @next/next/no-img-element */
"use client";

import { Dispatch, SetStateAction, useRef, useState } from "react";
import styles from "./FileUpload.module.scss";
import { Button } from "../button/Button";
import { FaTimes } from "react-icons/fa";

interface FileUploadProps {
  setFile: Dispatch<SetStateAction<File | null>>;
  handleUploadImage: () => void;
  handleRemoved: () => void;
}

export function FileUpload(props: FileUploadProps) {
  const { setFile, handleUploadImage, handleRemoved } = props;
  const fileInputRef = useRef<HTMLInputElement>(null);
  const [previewUrl, setPreviewUrl] = useState<string | null>(null);

  const handleFileAdd = () => {
    fileInputRef.current!.click();
  };

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files;
    if (files && files.length) {
      const file = files[0];

      const reader = new FileReader();
      reader.onloadend = () => {
        setPreviewUrl(reader.result as string);
      };
      reader.readAsDataURL(file);

      e.target.value = "";
      setFile(file);
      handleUploadImage();
    }
  };

  const handleRemoveImage = () => {
    setPreviewUrl(null);
    handleRemoved();
  };

  return (
    <>
      {previewUrl ? (
        <div className={styles.previewContainer}>
          <img src={previewUrl} alt="Preview" className={styles.preview} />
          <button onClick={handleRemoveImage} className={styles.closeButton}>
            <FaTimes />
          </button>
        </div>
      ) : (
        <label htmlFor="fileInput">
          <Button onClick={handleFileAdd}>画像をアップロード</Button>
          <input
            ref={fileInputRef}
            id="fileInput"
            type="file"
            accept="image/*"
            className={styles.inputHidden}
            onChange={handleFileChange}
          />
        </label>
      )}
    </>
  );
}
