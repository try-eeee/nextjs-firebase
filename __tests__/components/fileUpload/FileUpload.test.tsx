import React from "react";
import { render } from "@testing-library/react";
import { FileUpload } from "@/components/fileUpload/FileUpload";

describe("FileUpload UnitTest", () => {
  test("Snapshot Test Of FileUpload", () => {
    const { asFragment } = render(
      <FileUpload
        setFile={() => {}}
        handleUploadImage={() => {}}
        handleRemoved={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
