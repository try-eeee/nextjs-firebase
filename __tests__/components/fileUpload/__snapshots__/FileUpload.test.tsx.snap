// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload UnitTest Snapshot Test Of FileUpload 1`] = `
<DocumentFragment>
  <label
    for="fileInput"
  >
    <button
      class="button default size-S"
    >
      画像をアップロード
    </button>
    <input
      accept="image/*"
      class="inputHidden"
      id="fileInput"
      type="file"
    />
  </label>
</DocumentFragment>
`;
